File: web/frontend/src/services/salesInvoiceService.ts
import apiClient from './apiClient';

/**
 * فاتورة المبيعات (Sales Invoice)
 * API: GET/POST /sales/invoices, GET /sales/invoices/:id
 * تُنشأ من إذن الصرف أو أمر البيع.
 */

export interface SalesInvoiceItemDto {
    id?: number;
    salesInvoiceId?: number;
    issueItemId?: number;
    itemId: number;
    itemCode?: string;
    itemNameAr?: string;
    itemNameEn?: string;
    description?: string;
    quantity: number;
    unitId: number;
    unitNameAr?: string;
    unitPrice: number;
    unitCost?: number;
    discountPercentage?: number;
    discountAmount?: number;
    taxPercentage?: number;
    taxAmount?: number;
    totalPrice: number;
    grossProfit?: number;
}

export interface SalesInvoiceDto {
    id?: number;
    invoiceNumber?: string;
    invoiceDate: string;
    dueDate: string;
    salesOrderId?: number;
    soNumber?: string;
    issueNoteId?: number;
    issueNoteNumber?: string;
    customerId: number;
    customerNameAr?: string;
    customerCode?: string;
    salesRepId?: number;
    currency?: string;
    exchangeRate?: number;
    subTotal?: number;
    discountPercentage?: number;
    discountAmount?: number;
    taxAmount?: number;
    shippingCost?: number;
    totalAmount?: number;
    paidAmount?: number;
    remainingAmount?: number;
    status?: string;
    eInvoiceStatus?: string;
    eInvoiceUUID?: string;
    paymentTerms?: string;
    notes?: string;
    createdAt?: string;
    createdBy?: number;
    updatedAt?: string;
    updatedBy?: number;
    items: SalesInvoiceItemDto[];
}

const _api = '/sales/invoices';

// Legacy interface for backward compatibility with existing pages
export interface LegacySalesInvoiceItemDto {
    itemId: number;
    itemNameAr?: string;
    itemCode?: string;
    qty: number;
    unitId: number;
    unitNameAr?: string;
    unitPrice: number;
    discountPercent?: number;
    total?: number;
}

export interface LegacySalesInvoiceDto {
    id?: number;
    invoiceNumber?: string;
    invoiceDate: string;
    saleOrderId?: number;
    saleOrderNumber?: string;
    materialIssueId?: number;
    materialIssueNumber?: string;
    customerId: number;
    customerNameAr?: string;
    customerNameEn?: string;
    dueDate?: string;
    paymentTerms?: string;
    currency?: string;
    exchangeRate?: number;
    subTotal?: number;
    discountPercent?: number;
    discountAmount?: number;
    taxPercent?: number;
    taxAmount?: number;
    totalAmount?: number;
    paidAmount?: number;
    balanceAmount?: number;
    status?: string;
    notes?: string;
    items: LegacySalesInvoiceItemDto[];
}

// Mapping functions
const mapToLegacyItem = (item: SalesInvoiceItemDto): LegacySalesInvoiceItemDto => ({
    itemId: item.itemId,
    itemNameAr: item.itemNameAr,
    itemCode: item.itemCode,
    qty: item.quantity,
    unitId: item.unitId,
    unitNameAr: item.unitNameAr,
    unitPrice: item.unitPrice,
    discountPercent: item.discountPercentage,
    total: item.totalPrice,
});

const mapFromLegacyItem = (item: LegacySalesInvoiceItemDto): SalesInvoiceItemDto => ({
    itemId: item.itemId,
    itemNameAr: item.itemNameAr,
    itemCode: item.itemCode,
    quantity: item.qty,
    unitId: item.unitId,
    unitNameAr: item.unitNameAr,
    unitPrice: item.unitPrice,
    discountPercentage: item.discountPercent,
    discountAmount: 0,
    taxPercentage: 0,
    taxAmount: 0,
    totalPrice: item.total || (item.qty * item.unitPrice * (1 - (item.discountPercent || 0) / 100)),
});

const mapToLegacy = (dto: SalesInvoiceDto): LegacySalesInvoiceDto => ({
    id: dto.id,
    invoiceNumber: dto.invoiceNumber,
    invoiceDate: dto.invoiceDate,
    saleOrderId: dto.salesOrderId,
    saleOrderNumber: dto.soNumber,
    materialIssueId: dto.issueNoteId,
    materialIssueNumber: dto.issueNoteNumber,
    customerId: dto.customerId,
    customerNameAr: dto.customerNameAr,
    customerNameEn: undefined,
    dueDate: dto.dueDate,
    paymentTerms: dto.paymentTerms,
    currency: dto.currency,
    exchangeRate: dto.exchangeRate,
    subTotal: dto.subTotal,
    discountPercent: dto.discountPercentage,
    discountAmount: dto.discountAmount,
    taxPercent: undefined,
    taxAmount: dto.taxAmount,
    totalAmount: dto.totalAmount,
    paidAmount: dto.paidAmount,
    balanceAmount: dto.remainingAmount,
    status: dto.status,
    notes: dto.notes,
    items: dto.items.map(mapToLegacyItem),
});

const mapFromLegacy = (dto: LegacySalesInvoiceDto): SalesInvoiceDto => ({
    id: dto.id,
    invoiceNumber: dto.invoiceNumber,
    invoiceDate: dto.invoiceDate,
    dueDate: dto.dueDate || new Date().toISOString().split('T')[0],
    salesOrderId: dto.saleOrderId,
    soNumber: dto.saleOrderNumber,
    issueNoteId: dto.materialIssueId,
    issueNoteNumber: dto.materialIssueNumber,
    customerId: dto.customerId,
    customerNameAr: dto.customerNameAr,
    customerCode: undefined,
    salesRepId: undefined,
    currency: dto.currency,
    exchangeRate: dto.exchangeRate,
    subTotal: dto.subTotal,
    discountPercentage: dto.discountPercent,
    discountAmount: dto.discountAmount,
    taxAmount: dto.taxAmount,
    shippingCost: 0,
    totalAmount: dto.totalAmount,
    paidAmount: dto.paidAmount,
    remainingAmount: dto.balanceAmount,
    status: dto.status,
    paymentTerms: dto.paymentTerms,
    notes: dto.notes,
    items: dto.items.map(mapFromLegacyItem),
});

export const salesInvoiceService = {
    getAll: async (): Promise<SalesInvoiceDto[]> => {
        try {
            const res = await apiClient.get<{ data?: SalesInvoiceDto[] }>(_api);
            return (res.data as any)?.data ?? [];
        } catch {
            return [];
        }
    },
    getById: async (id: number): Promise<SalesInvoiceDto | LegacySalesInvoiceDto | null> => {
        try {
            const res = await apiClient.get<{ data?: SalesInvoiceDto }>(`${_api}/${id}`);
            const dto = (res.data as any)?.data ?? null;
            // Return in legacy format for backward compatibility
            return dto ? mapToLegacy(dto) : null;
        } catch {
            return null;
        }
    },
    create: async (dto: SalesInvoiceDto | LegacySalesInvoiceDto): Promise<SalesInvoiceDto | null> => {
        const payload = 'qty' in (dto.items?.[0] || {}) ? mapFromLegacy(dto as LegacySalesInvoiceDto) : dto as SalesInvoiceDto;
        const res = await apiClient.post<{ data?: SalesInvoiceDto }>(_api, payload);
        return (res.data as any)?.data ?? null;
    },
    update: async (id: number, dto: SalesInvoiceDto | LegacySalesInvoiceDto): Promise<SalesInvoiceDto | null> => {
        try {
            const payload = 'qty' in (dto.items?.[0] || {}) ? mapFromLegacy(dto as LegacySalesInvoiceDto) : dto as SalesInvoiceDto;
            const res = await apiClient.put<{ data?: SalesInvoiceDto }>(`${_api}/${id}`, payload);
            return (res.data as any)?.data ?? null;
        } catch {
            return null;
        }
    },
    delete: async (id: number): Promise<boolean> => {
        try {
            await apiClient.delete(`${_api}/${id}`);
            return true;
        } catch {
            return false;
        }
    },
    // Legacy methods that return legacy format
    getAllLegacy: async (): Promise<LegacySalesInvoiceDto[]> => {
        const dtos = await salesInvoiceService.getAll();
        return dtos.map(mapToLegacy);
    },
    getByIdLegacy: async (id: number): Promise<LegacySalesInvoiceDto | null> => {
        const dto = await salesInvoiceService.getById(id);
        return dto ? mapToLegacy(dto) : null;
    },
};

File: web/backend/src/main/java/com/rasras/erp/sales/SalesInvoiceService.java
package com.rasras.erp.sales;

import com.rasras.erp.crm.Customer;
import com.rasras.erp.crm.CustomerRepository;
import com.rasras.erp.inventory.Item;
import com.rasras.erp.inventory.ItemRepository;
import com.rasras.erp.inventory.UnitOfMeasure;
import com.rasras.erp.inventory.UnitRepository;
import com.rasras.erp.approval.ApprovalService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class SalesInvoiceService {

    private final SalesInvoiceRepository invoiceRepository;
    private final SalesOrderRepository orderRepository;
    private final CustomerRepository customerRepository;
    private final ItemRepository itemRepository;
    private final UnitRepository unitRepository;
    private final ApprovalService approvalService;

    @Transactional(readOnly = true)
    public List<SalesInvoiceDto> getAllInvoices() {
        return invoiceRepository.findAll().stream()
                .map(this::mapToDto)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public SalesInvoiceDto getInvoiceById(Integer id) {
        return invoiceRepository.findById(id)
                .map(this::mapToDto)
                .orElseThrow(() -> new RuntimeException("Sales Invoice not found"));
    }

    @Transactional
    public SalesInvoiceDto createInvoice(SalesInvoiceDto dto) {
        SalesInvoice invoice = new SalesInvoice();
        invoice.setInvoiceNumber(generateInvoiceNumber());
        invoice.setInvoiceDate(
                dto.getInvoiceDate() != null ? dto.getInvoiceDate().atStartOfDay() : LocalDateTime.now());
        invoice.setDueDate(dto.getDueDate() != null ? dto.getDueDate() : LocalDate.now().plusDays(30));

        if (dto.getSalesOrderId() != null) {
            SalesOrder order = orderRepository.findById(dto.getSalesOrderId())
                    .orElse(null);
            invoice.setSalesOrder(order);
        }

        invoice.setIssueNoteId(dto.getIssueNoteId());

        Customer customer = customerRepository.findById(dto.getCustomerId())
                .orElseThrow(() -> new RuntimeException("Customer not found"));
        invoice.setCustomer(customer);

        invoice.setSalesRepId(dto.getSalesRepId());
        invoice.setCurrency(dto.getCurrency() != null ? dto.getCurrency() : "EGP");
        invoice.setExchangeRate(dto.getExchangeRate() != null ? dto.getExchangeRate() : BigDecimal.ONE);
        invoice.setSubTotal(dto.getSubTotal() != null ? dto.getSubTotal() : BigDecimal.ZERO);
        invoice.setDiscountPercentage(
                dto.getDiscountPercentage() != null ? dto.getDiscountPercentage() : BigDecimal.ZERO);
        invoice.setDiscountAmount(dto.getDiscountAmount() != null ? dto.getDiscountAmount() : BigDecimal.ZERO);
        invoice.setTaxAmount(dto.getTaxAmount() != null ? dto.getTaxAmount() : BigDecimal.ZERO);
        invoice.setShippingCost(dto.getShippingCost() != null ? dto.getShippingCost() : BigDecimal.ZERO);
        invoice.setTotalAmount(dto.getTotalAmount() != null ? dto.getTotalAmount() : BigDecimal.ZERO);
        invoice.setPaidAmount(dto.getPaidAmount() != null ? dto.getPaidAmount() : BigDecimal.ZERO);
        invoice.setStatus(dto.getStatus() != null ? dto.getStatus() : "Draft");
        invoice.setApprovalStatus(dto.getApprovalStatus() != null ? dto.getApprovalStatus() : "Pending");
        invoice.setPaymentTerms(dto.getPaymentTerms());
        invoice.setNotes(dto.getNotes());
        invoice.setCreatedBy(dto.getCreatedBy() != null ? dto.getCreatedBy() : 1);

        if (dto.getItems() != null && !dto.getItems().isEmpty()) {
            List<SalesInvoiceItem> items = new ArrayList<>();
            for (SalesInvoiceItemDto itemDto : dto.getItems()) {
                SalesInvoiceItem item = mapItemToEntity(itemDto, invoice);
                items.add(item);
            }
            invoice.setItems(items);
        }

        SalesInvoice saved = invoiceRepository.save(invoice);
        return mapToDto(saved);
    }

    @Transactional
    public SalesInvoiceDto updateInvoice(Integer id, SalesInvoiceDto dto) {
        SalesInvoice invoice = invoiceRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Sales Invoice not found"));

        if (!"Draft".equals(invoice.getStatus()) && !"Rejected".equals(invoice.getStatus())) {
            throw new RuntimeException("Cannot edit invoice that is not in Draft or Rejected status");
        }

        // Reset approval status if rejected and being edited
        if ("Rejected".equals(invoice.getStatus())) {
            invoice.setApprovalStatus("Pending");
        }

        invoice.setDueDate(dto.getDueDate());
        invoice.setSubTotal(dto.getSubTotal());
        invoice.setDiscountPercentage(dto.getDiscountPercentage());
        invoice.setDiscountAmount(dto.getDiscountAmount());
        invoice.setTaxAmount(dto.getTaxAmount());
        invoice.setShippingCost(dto.getShippingCost());
        invoice.setTotalAmount(dto.getTotalAmount());
        invoice.setPaymentTerms(dto.getPaymentTerms());
        invoice.setNotes(dto.getNotes());
        invoice.setUpdatedBy(dto.getUpdatedBy() != null ? dto.getUpdatedBy() : 1);

        if (invoice.getItems() != null) {
            invoice.getItems().clear();
        } else {
            invoice.setItems(new ArrayList<>());
        }

        if (dto.getItems() != null) {
            for (SalesInvoiceItemDto itemDto : dto.getItems()) {
                SalesInvoiceItem item = mapItemToEntity(itemDto, invoice);
                invoice.getItems().add(item);
            }
        }

        return mapToDto(invoiceRepository.save(invoice));
    }

    @Transactional
    public SalesInvoiceDto submitForApproval(Integer id) {
        SalesInvoice invoice = invoiceRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Sales Invoice not found"));

        if (!"Draft".equals(invoice.getStatus()) && !"Rejected".equals(invoice.getStatus())) {
            throw new RuntimeException("Sales Invoice must be in Draft or Rejected status to submit");
        }

        invoice.setStatus("Pending");
        SalesInvoice saved = invoiceRepository.save(invoice);

        // Initiate approval workflow
        approvalService.initiateApproval("INVOICE_APPROVAL", "SalesInvoice", saved.getId(),
                saved.getInvoiceNumber(), saved.getCreatedBy() != null ? saved.getCreatedBy() : 1,
                saved.getTotalAmount());

        return mapToDto(saved);
    }

    @Transactional
    public void deleteInvoice(Integer id) {
        SalesInvoice invoice = invoiceRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Sales Invoice not found"));

        if (!"Draft".equals(invoice.getStatus()) && !"Pending".equals(invoice.getStatus()) && !"Rejected".equals(invoice.getStatus())) {
            throw new RuntimeException("Cannot delete invoice that is not in Draft, Pending, or Rejected status");
        }

        invoiceRepository.delete(invoice);
    }

    private String generateInvoiceNumber() {
        return "INV-" + System.currentTimeMillis();
    }

    private SalesInvoiceDto mapToDto(SalesInvoice invoice) {
        SalesInvoiceDto dto = SalesInvoiceDto.builder()
                .id(invoice.getId())
                .invoiceNumber(invoice.getInvoiceNumber())
                .invoiceDate(invoice.getInvoiceDate() != null ? invoice.getInvoiceDate().toLocalDate() : null)
                .dueDate(invoice.getDueDate())
                .issueNoteId(invoice.getIssueNoteId())
                .customerId(invoice.getCustomer().getId())
                .customerNameAr(invoice.getCustomer().getCustomerNameAr())
                .customerCode(invoice.getCustomer().getCustomerCode())
                .salesRepId(invoice.getSalesRepId())
                .currency(invoice.getCurrency())
                .exchangeRate(invoice.getExchangeRate())
                .subTotal(invoice.getSubTotal())
                .discountPercentage(invoice.getDiscountPercentage())
                .discountAmount(invoice.getDiscountAmount())
                .taxAmount(invoice.getTaxAmount())
                .shippingCost(invoice.getShippingCost())
                .totalAmount(invoice.getTotalAmount())
                .paidAmount(invoice.getPaidAmount())
                .remainingAmount(invoice.getTotalAmount().subtract(invoice.getPaidAmount()))
                .status(invoice.getStatus())
                .approvalStatus(invoice.getApprovalStatus())
                .eInvoiceStatus(invoice.getEInvoiceStatus())
                .eInvoiceUUID(invoice.getEInvoiceUUID())
                .paymentTerms(invoice.getPaymentTerms())
                .notes(invoice.getNotes())
                .createdAt(invoice.getCreatedAt())
                .createdBy(invoice.getCreatedBy())
                .updatedAt(invoice.getUpdatedAt())
                .updatedBy(invoice.getUpdatedBy())
                .build();

        if (invoice.getSalesOrder() != null) {
            dto.setSalesOrderId(invoice.getSalesOrder().getId());
            dto.setSoNumber(invoice.getSalesOrder().getSoNumber());
        }

        if (invoice.getItems() != null) {
            dto.setItems(invoice.getItems().stream()
                    .map(this::mapItemToDto)
                    .collect(Collectors.toList()));
        } else {
            dto.setItems(new ArrayList<>());
        }

        return dto;
    }

    private SalesInvoiceItemDto mapItemToDto(SalesInvoiceItem item) {
        return SalesInvoiceItemDto.builder()
                .id(item.getId())
                .salesInvoiceId(item.getSalesInvoice().getId())
                .issueItemId(item.getIssueItemId())
                .itemId(item.getItem().getId())
                .itemCode(item.getItem().getItemCode())
                .itemNameAr(item.getItem().getItemNameAr())
                .itemNameEn(item.getItem().getItemNameEn())
                .description(item.getDescription())
                .quantity(item.getQuantity())
                .unitId(item.getUnit().getId())
                .unitNameAr(item.getUnit().getUnitNameAr())
                .unitPrice(item.getUnitPrice())
                .unitCost(item.getUnitCost())
                .discountPercentage(item.getDiscountPercentage())
                .discountAmount(item.getDiscountAmount())
                .taxPercentage(item.getTaxPercentage())
                .taxAmount(item.getTaxAmount())
                .totalPrice(item.getTotalPrice())
                .grossProfit(item.getTotalPrice().subtract(
                        item.getQuantity().multiply(item.getUnitCost() != null ? item.getUnitCost() : BigDecimal.ZERO)))
                .build();
    }

    private SalesInvoiceItem mapItemToEntity(SalesInvoiceItemDto dto, SalesInvoice invoice) {
        Item item = itemRepository.findById(dto.getItemId())
                .orElseThrow(() -> new RuntimeException("Item not found"));

        UnitOfMeasure unit = unitRepository.findById(dto.getUnitId())
                .orElseThrow(() -> new RuntimeException("Unit not found"));

        return SalesInvoiceItem.builder()
                .salesInvoice(invoice)
                .issueItemId(dto.getIssueItemId())
                .item(item)
                .description(dto.getDescription())
                .quantity(dto.getQuantity())
                .unit(unit)
                .unitPrice(dto.getUnitPrice())
                .unitCost(dto.getUnitCost())
                .discountPercentage(dto.getDiscountPercentage() != null ? dto.getDiscountPercentage() : BigDecimal.ZERO)
                .discountAmount(dto.getDiscountAmount() != null ? dto.getDiscountAmount() : BigDecimal.ZERO)
                .taxPercentage(dto.getTaxPercentage() != null ? dto.getTaxPercentage() : BigDecimal.ZERO)
                .taxAmount(dto.getTaxAmount() != null ? dto.getTaxAmount() : BigDecimal.ZERO)
                .totalPrice(dto.getTotalPrice())
                .build();
    }
}
